### DISCUSSION TITLE
Why up_for_retry takes worker slot?

### DISCUSSION THREAD

**Discussion by workingDice** [apache/airflow#37986] [Q&A] [answered]
Since up_for_retry is just waiting there for a certain retry delay, shouldn't it also pause execution when idle, resume when condition changes?

**Comment by potiuk** [comment, marked as answer]
This is what Deferrable Operators do https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/deferring.html. If you understand how running programs work you cannot just `vanish` a process that is waiting. When the process runs and does `sleep` (which is what  `up_for_retry` does) the process cannot magically release all it's resources (memory, running process, sockets etc.). The process is still there - the only thing it does not use when waiting is CPU - but all other resources are still occupied. You cannot "just" release all the memory and process as - for example - it keeps the state of your process and just "exiting" the process (which what you effectively ask for) would loose all that information, also you need something that will effectively resume your process after the time passes (it will not magically reappear out of the blue - you need to store somewhere the fact that after the time passes, your task will have to resume from the place it left off.

This is what Deferrable Operators solve. They work in the way  that when your task decide to Defer itself, the process state is serialized and stored in the Database, then Triggerer will add such defferred process to an asyncio event loop where all deferred processes are running Triggers, which will fire the condition the trigger is waiting for (it might be time, or message coming on opened socket, or any other asyncio - compatible event). Then such trigger will deserialize state of the process and executor will restart the deferred process after deserializing and restoring state of the task to the moment it was when it was deferred.

**Reply by workingDice** [reply]
Thank you Jarek, I really appreciate your detailed explanation.

**Reply by workingDice** [reply]
Hi Jarek, a followed up question:
From my understanding, `retries` and `retry_delay` should be small. And if we really want to give some tasks a very long retry delay, we should achieve it using deferrable operators instead. Am I right?

**Reply by potiuk** [reply]
More or less, yes.

**Reply by workingDice** [reply]
Thank you! I think it would be very helpful if such instruction or warning could be added to the airflow documentation.

**Reply by potiuk** [reply]
Absolutely. Now when you got it explained - it's alll yours to make it. 

It's super easy. Open a page where you think such information should be added, click at the right bottom "Suggest a change on this page" and PR will be opened for you automatically and you will be able to edit the source of the documentation this way - straight there in your GitHub UI. Without you having to setup anything. 

Your PR will be reviewed (so you won't have to afraid contributing), as any other and you can join > 2800 contributors to Airflow - many of them users like you who found out that something could be changed - documentation updated, bug fixed, feature added.

Can we count on your contribution there @workingDice now when you got everything laid out so nicely?

**Reply by potiuk** [reply]
Actually - you are probably the best person in the world to do it. You needed it, you looked for it and have not found it (so you know where to add it because you know where you looked for it), you got it fresh in your memory after being explained, and you can phrase it all in the way that people who have similar background to yours would understand and appreciate (unlike maintainers who have too many assumptions in their head).

**Reply by workingDice** [reply]
lol sure, I'll do it later this week, thank you
